refactor: convert `createColorHOC` to functional component.

USERSATOSHI opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->
See #22890
This PR refactors `createColorHOC` from class based component syntax to newer functional component syntax.

<!-- In a few words, what is the PR actually doing? -->

## Why?
Class based components are no longer the recommended approach & instead functional syntax with hooks is more recommended.
The code becomes more readable & easier to maintain.

## How?
- Converted the class-based HOC to a functional component using React hooks.
- Used `useState` to manage internal derived color state.
- Replaced `getDerivedStateFromProps` with a `useEffect` that recalculates color state when attributes or colors change.
- Created color setter functions using `useMemo` to match the dynamic method generation in the original class.
- Implemented `colorUtils.getMostReadableColor` with `useCallback` to preserve referential integrity.

## Testing Instructions
- To test this function, use `Featured Post Image` block which uses this internally.
- Expect same behaviour as before

## Screenshots or screencast <!-- if applicable -->

<!-- If you would like to upload screenshots, feel free to use the table below when it is useful to show the difference between before and after the change. -->

USERSATOSHI committed:
refactor: convert class component to functional component with hooks in `createColorHOC`

USERSATOSHI committed:
fix: add lazy state init before initial render

USERSATOSHI committed:
Merge branch 'trunk' into try/refactor-with-color-functional-component

USERSATOSHI committed:
Merge branch 'trunk' into try/refactor-with-color-functional-component

USERSATOSHI committed:
Merge branch 'trunk' into try/refactor-with-color-functional-component

USERSATOSHI committed:
Merge branch 'trunk' into try/refactor-with-color-functional-component

USERSATOSHI committed:
Merge branch 'trunk' into try/refactor-with-color-functional-component

USERSATOSHI committed:
Merge branch 'trunk' into try/refactor-with-color-functional-component

USERSATOSHI committed:
Merge branch 'WordPress:trunk' into try/refactor-with-color-functional-component

USERSATOSHI committed:
Merge branch 'trunk' into try/refactor-with-color-functional-component

USERSATOSHI committed:
merge trunk

USERSATOSHI committed:
Merge branch 'try/refactor-with-color-functional-component' of https://github.com/usersatoshi/gutenberg into try/refactor-with-color-functional-component

USERSATOSHI committed:
Merge branch 'trunk' into try/refactor-with-color-functional-component

USERSATOSHI commented:
<img width="1365" height="531" alt="image" src="https://github.com/user-attachments/assets/8dcdb38b-f6fb-4887-9f6f-62c3f9fe94b7" />
Seems like there is some duplicate triggers happening. I am checking that right now!

USERSATOSHI committed:
fix: compute derived color state during render, not in effect

USERSATOSHI committed:
test: assert for only first arg

Mamaduka committed:
Merge branch 'trunk' into try/refactor-with-color-functional-component

Mamaduka committed:
More cleanup

tyxla commented on packages/block-editor/src/components/colors/with-colors.js:
```diff
@@ -94,122 +94,90 @@ function createColorHOC( colorTypes, withColorPalette ) {
+				const colorState = useMemo(
+					() => computeDerivedColorState( attributes, colors ),
+					[ attributes, colors ]
+				);
```
Minor: the new code memoizes the whole `colorState` on `[ attributes, colors ]`. Since a block's attributes object gets a new reference on every attribute edit (including edits unrelated to color), any attribute change now rebuilds all color objects. In contrast, the old code preserved identity for colors whose resolved value didn't change. 

This doesn't feel like a blocker to me, but something to highlight as a minor change from before, and something we might have done unintentionally.

jsnajdr replied:
Yes, there are some `getDerivedStateFromProps` scenarios that modern functional components with hooks can't reproduce easily. Computing the derived state has access to the previous state, but in a functional component we don't have that.

It could be solved in a followup with some smart memoization.

tyxla commented on packages/block-editor/src/components/colors/test/with-colors.js:
```diff
@@ -22,23 +22,21 @@ describe( 'createCustomColorsHOC', () => {
-			} ),
-			expect.anything()
-		);
+		const props = BaseComponent.mock.calls[ 0 ][ 0 ];
```
Why do we need to pick the exact call instead of just keeping `toHaveBeenCalledWith()`?

USERSATOSHI replied:
For me, calls[0][0] + toMatchObject() is more readable here. It reads like a plain object compared to nested expect.objectContaining(). 

It also keeps the props reference handy if we later need to assert on individual props or invoke callbacks, without changing the approach. 
Since the component only renders once, there's no ambiguity about which call we're inspecting.

tyxla replied:
My $0.02: the testing framework has an API, and to me that's more readable than inspecting the internals. 

This isn't a blocker though, just a personal preference.

jsnajdr replied:
It's true that `toMatchObject` and `expect.objectContaining` behave differently when it comes to partially matching deep objects. And unfortunately they are not interchangeable, one is a "symmetric" matcher, the other one is "asymmetric".

I'd also prefer the `.toHaveBeenCalledWith` syntax, if it's possible.

tyxla reviewed (commented):
Nice refactor! Nothing major from my end. Just a couple of little suggestions/comments.

jsnajdr commented on packages/block-editor/src/components/colors/with-colors.js:
```diff
@@ -94,122 +94,90 @@ function createColorHOC( colorTypes, withColorPalette ) {
+				const colorState = useMemo(
+					() => computeDerivedColorState( attributes, colors ),
+					[ attributes, colors ]
+				);
```
`colorState` and `computeDerivedColorState` are no longer good variable/function names, because we're not dealing with state and derived state any more. `colorState` can be called `colorValues`, analogous to existing `colorSetters`.

I wouldn't be afraid to merge the `colorState` and `colorGetters` computation into one `useMemo` and one loop through `colorMap`. For each color, we have one value (getter) and one setter.

Instead of `reduce`, more popular today would be `Object.fromEntries( Object.entries( colorMap ).flatMap )`. The `flatMap` callback returns array of two entries, getter and setter.

jsnajdr commented on packages/block-editor/src/components/colors/with-colors.js:
```diff
@@ -94,122 +94,90 @@ function createColorHOC( colorTypes, withColorPalette ) {
+					<WrappedComponent
+						{ ...{
+							...props,
+							colors: undefined,
```
Instead of resetting `colors`, we could exclude them from the passed props:
```jsx
const { colors, ...passedProps } = props;
const { attributes, setAttributes } = passedProps;

<WrappedComponent { ...passedProps } { ...colorProps } />
```

jsnajdr reviewed (approved):
I think this can be merged as-is. All feedback can be solved in a followup PR.

USERSATOSHI committed:
refactor(with-colors): merge color getter/setter into single useMemo with identity preservation

- Replace `colors: undefined` spread with destructuring colors from props
- Merge color value and setter computation into one useMemo via
  Object.fromEntries(Object.entries().flatMap())
- Preserve color object identity across unrelated attribute edits via useRef
- Use toHaveBeenCalledWith() in test instead of calls[0][0]

USERSATOSHI committed:
refactor(with-colors): merge color getter/setter into single useMemo with identity preservation

- Replace `colors: undefined` spread with destructuring colors from props
- Merge color value and setter computation into one useMemo via
  Object.fromEntries(Object.entries().flatMap())
- Preserve color object identity across unrelated attribute edits via useRef
- Use toHaveBeenCalledWith() in test instead of calls[0][0]

USERSATOSHI committed:
chore: fix lint error

Mamaduka commented on packages/block-editor/src/components/colors/with-colors.js:
```diff
@@ -1,4 +1,7 @@
-import { useMemo, Component } from '@wordpress/element';
+/**
+ * WordPress dependencies
+ */
```
I think this was the reason for the conflict and now the errors - Unexpected dependency group comment block.

USERSATOSHI replied:
removed it, ci now passes. we should be good to merge the PR
